rust/agama-network/src/dbus/interfaces.rs

```diff
@@ -0,0 +1,402 @@
+    pub fn new(objects: Arc<Mutex<ObjectsRegistry>>, actions: Sender<Action>) -> Self {
+        Self {
+            objects,
+            actions: Arc::new(Mutex::new(actions)),
+        }
+    }
+}
+
+#[dbus_interface(name = "org.opensuse.Agama.Network1.Connections")]
+impl Connections {
+    /// Returns the D-Bus paths of the network connections.
+    pub fn get_connections(&self) -> Vec<ObjectPath> {
+        let objects = self.objects.lock();
+        objects
+            .connections_paths()
+            .iter()
+            .filter_map(|c| ObjectPath::try_from(c.clone()).ok())
+            .collect()
+    }
+
+    /// Adds a new network connection.
+    ///
+    /// * `name`: connection name.
+    /// * `ty`: connection type (see [crate::model::DeviceType]).
+    pub async fn add_connection(&mut self, name: String, ty: u8) -> zbus::fdo::Result<()> {
+        let actions = self.actions.lock();
+        actions
+            .send(Action::AddConnection(name, ty.try_into()?))
+            .unwrap();
+        Ok(())
+    }
+
+    /// Removes a network connection.
+    ///
+    /// * `uuid`: connection UUID..
+    pub async fn remove_connection(&mut self, uuid: &str) -> zbus::fdo::Result<()> {
+        let actions = self.actions.lock();
+        let uuid =
+            Uuid::parse_str(uuid).map_err(|_| NetworkStateError::InvalidUuid(uuid.to_string()))?;
+        actions.send(Action::RemoveConnection(uuid)).unwrap();
+        Ok(())
+    }
+
+    /// Applies the network configuration.
+    ///
+    /// It includes adding, updating and removing connections as needed.
+    pub async fn apply(&self) -> zbus::fdo::Result<()> {
+        let actions = self.actions.lock();
+        actions.send(Action::Apply).unwrap();
+        Ok(())
+    }
+}
+
+/// D-Bus interface for a network connection
+///
+/// It offers an API to query a connection.
+pub struct Connection {
+    connection: Arc<Mutex<NetworkConnection>>,
+}
+
+impl Connection {
+    /// Creates a Connection interface object.
+    ///
+    /// * `UUID`: Connection UUID.
+    pub fn new(connection: Arc<Mutex<NetworkConnection>>) -> Self {
+        Self { connection }
+    }
+
+    /// Returns the underlying connection.
+    fn get_connection(&self) -> MutexGuard<NetworkConnection> {
+        self.connection.lock()
+    }
+}
+
+#[dbus_interface(name = "org.opensuse.Agama.Network1.Connection")]
+impl Connection {
+    #[dbus_interface(property)]
+    pub fn id(&self) -> String {
+        self.get_connection().id().to_string()
+    }
+
+    #[dbus_interface(property, name = "UUID")]
+    pub fn uuid(&self) -> String {
+        self.get_connection().uuid().to_string()
+    }
+}
+
+/// D-Bus interface for IPv4 settings
+pub struct Ipv4 {
+    actions: Arc<Mutex<Sender<Action>>>,
+    connection: Arc<Mutex<NetworkConnection>>,
+}
+
+impl Ipv4 {
+    /// Creates an IPv4 interface object.
+    ///
+    /// * `actions`: sending-half of a channel to send actions.
```

Comment:
I do not get what means `sending-half of a channel`

Reply:
[sending half](https://doc.rust-lang.org/std/sync/mpsc/struct.Sender.html) and [receiving half](https://doc.rust-lang.org/std/sync/mpsc/struct.Receiver.html) are the the two parts of a channel.

Reply:
yeah, after reading about channels it makes perfect sense...I just was not familiar (or have association ) with term channel